biogeochem/EDCanopyStructureMod.F90

```diff
@@ -906,110 +984,164 @@ subroutine PromoteIntoLayer(currentSite,currentPatch,i_lyr)
-                                         remainder_area_hold * currentCohort%prom_weight/sumweights_old, &
-                                         currentCohort%c_area-currentCohort%prom_weight)
-
-
-                           ! Reduce the remainder_area
-                           remainder_area = remainder_area - cc_gain
-                           
-                           ! Update this cohort's promotion
-                           currentCohort%prom_weight = currentCohort%prom_weight + cc_gain
-                           
-                           ! Update the sum of weights for cohorts where exceedance can
-                           ! still be spread to
-                           if( currentCohort%prom_weight < (currentCohort%c_area-nearzero) ) then
-                              sumweights = sumweights + currentCohort%prom_weight
+                  do while (associated(currentCohort))
+                     if(currentCohort%canopy_layer  ==  (i_lyr+1) ) then 
+                        currentCohort%prom_weight = currentCohort%c_area * &
+                              currentCohort%prom_weight * scale_factor
+                        
+                        if(debug)then
+                           if((currentCohort%prom_weight > &
+                                 (currentCohort%c_area+area_target_precision)) .or. &
+                                 (currentCohort%prom_weight < 0._r8)  ) then
+                               write(fates_log(),*) 'promotion area too big (1)'
+                               write(fates_log(),*) 'currentCohort%c_area: ',currentCohort%c_area
+                               write(fates_log(),*) 'currentCohort%prom_weight: ', &
+                                     currentCohort%prom_weight
+                               write(fates_log(),*) 'excess: ', &
+                                     currentCohort%prom_weight - currentCohort%c_area
+                               call endrun(msg=errMsg(sourcefile, __LINE__))
                            end if
-
-                        end if
-                     end if
-                     currentCohort => currentCohort%shorter 
-                  end do
-                  sumweights_old = sumweights
+                       end if
+                        
+                     endif
+                     currentCohort => currentCohort%shorter      
+                  enddo
                   
-                  exceedance_counter = exceedance_counter + 1
-                  if( exceedance_counter > 100 ) then
-                     write(fates_log(),*) 'It is taking a long time to distribute promotion exceedance'
-                     write(fates_log(),*) '(ie greater than c_area) to neighbors... exiting'
-                     call endrun(msg=errMsg(sourcefile, __LINE__))
-                  end if
+               else
                   
-               end do
+                  ! Non-trivial case, at least 1 cohort's demotion
```

Comment:
Comment typo: should read promotion, not demotion, correct?

Reply:
yup, good catch